### PR TITLE
Separate backward compatible test suite

### PR DESCRIPTION
**Pull request by mdumandag**
This PR separates the integration tests further into a suite
called backward compatible. Tests in this suite are meant to be
run against all released Node.js 4.x clients.

To achieve this, I made sure that either we :

- use public APIs in tests
- disable tests conditionally based on the server/client version if the
public API is not available on certain versions
- use private APIs that exist in all client versions (and likely to not change)
- conditionally use some private/public APIs based on the availability
of some parts

Also, increased the client version in the package.json to 4.2 to disable some
newly added tests conditionally.

And finally, changed the RestValue test a little bit (the idea is the same) to
make it pass on the Github Actions.

### CODE REVIEW

**Comment by srknzl** [comment]
To be sure, we have backward compatible tests just for integrations tests, right?

**Comment by mdumandag** [comment]
Yes, it is used with specific versions of the Hazelcast IMDG, so running unit tests does not make sense

File: test/integration/backward_compatible/map/MapProxyTest.js [hunk 1]
```diff
@@ -431,7 +436,7 @@ describe('MapProxyTest', function () {
                         expect(entry.expirationTime.isZero()).to.be.false;
                         expect(entry.hits.isZero()).to.be.false;
                         expect(entry.lastAccessTime.isZero()).to.be.false;
-                        expect(entry.lastStoredTime.isZero()).to.be.true;
+                        expect(entry.lastStoredTime.isZero()).to.be.equal(!TestUtil.isServerVersionAtLeast(client, '4.2'));
```
Comment (srknzl) [inline test/integration/backward_compatible/map/MapProxyTest.js:439]: What is different in 4.2 regarding `lastStoredTime.isZero`?

File: test/integration/backward_compatible/map/MapProxyTest.js [hunk 2]
```diff
@@ -355,6 +358,7 @@ describe('MapProxyTest', function () {
             });
 
             it('putTransient_with_maxIdle', async function () {
+                TestUtil.markClientVersionAtLeast(this, '4.1');
```
Comment (srknzl) [inline test/integration/backward_compatible/map/MapProxyTest.js:361]: I wonder the reason why clients older than 4.1 do not support these tests

File: test/integration/backward_compatible/map/NearCachedMapTest.js [hunk 3]
```diff
@@ -246,6 +246,7 @@ describe('NearCachedMapTest', function () {
             });
 
             it('setTtl invalidates the entry', async function () {
+                TestUtil.markClientVersionAtLeast(this, '4.1');
```
Comment (srknzl) [inline test/integration/backward_compatible/map/NearCachedMapTest.js:249]: Could you explain this as well

File: test/integration/backward_compatible/nearcache/MigratedDataTest.js [hunk 4]
```diff
@@ -33,10 +33,9 @@ describe('MigratedDataTest', function () {
 
     async function waitForPartitionTableEvent(partitionService) {
         const deferred = deferredPromise();
-        const expectedPartitionCount = partitionService.partitionCount;
 
         function checkPartitionTable(remainingTries) {
-            if (partitionService.partitionTable.partitions.size === expectedPartitionCount) {
+            if (partitionService.getPartitionOwner(0) != null) {
```
Comment (srknzl) [inline test/integration/backward_compatible/nearcache/MigratedDataTest.js:38]: Why this change is required? I guess getting the owner of partition 0 do not check if we get all partitions .

File: test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (srknzl) [inline test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml:1]: Is this file ever used ? If so, for what ?

File: test/integration/backward_compatible/serialization/DefaultSerializersLiveTest.js [hunk 6]
```diff
@@ -154,6 +155,7 @@ describe('DefaultSerializersLiveTest', function () {
     });
 
     it('should deserialize ArrayList', async function () {
+        TestUtil.markClientVersionAtLeast(this, '4.2');
```
Comment (srknzl) [inline test/integration/backward_compatible/serialization/DefaultSerializersLiveTest.js:158]: I guess these are something new

File: test/integration/backward_compatible/statistics/StatisticsTest.js [hunk 7]
```diff
@@ -97,7 +97,7 @@ describe('StatisticsTest (default period)', function () {
 
     function contains(base, search) {
         const firstIndex = base.indexOf(search);
-        return firstIndex > -1 && firstIndex == base.lastIndexOf(search);
+        return firstIndex > -1 && firstIndex === base.lastIndexOf(search);
```
Comment (srknzl) [inline test/integration/backward_compatible/statistics/StatisticsTest.js:100]: Looking at this code, I feel like its name can be containsOnce. We can do this change later(not in this pr) though

File: test/integration/javaclasses/PrefixFilter.js [hunk 8]
```diff
@@ -15,18 +15,20 @@
...
+
 function PrefixFilter(prefix) {
     this.prefix = prefix;
     this.factoryId = 66;
     this.classId = 4;
 }
 
 PrefixFilter.prototype.readData = function (inp) {
-    this.prefix = inp.readString();
+    this.prefix = TestUtil.readStringFromInput(inp);
```
Comment (srknzl) [inline test/integration/javaclasses/PrefixFilter.js:27]: Why change these other than the ones in backward_compatible folder

File: test/unit/serialization/PortableObjects.js [hunk 9]
```diff
@@ -25,13 +27,13 @@ class InnerPortable {
     }
 
     readPortable(reader) {
-        this.p1 = reader.readString('p1');
-        this.p2 = reader.readString('p2');
+        this.p1 = TestUtil.readStringFromReader(reader, 'p1');
```
Comment (srknzl) [inline test/unit/serialization/PortableObjects.js:30]: Same, why change these as these are not in backward_compatible folder

File: test/integration/backward_compatible/rest_value/RestValueTest.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        member = await RC.startMember(cluster.id);
+        client = await Client.newHazelcastClient({ clusterName: cluster.id });
+    });
+
+    after(async function () {
+        await client.shutdown();
+        return RC.terminateCluster(cluster.id);
+    });
+
+    it('client should receive REST events from server as RestValue', async function () {
```
Comment (srknzl) [inline test/integration/backward_compatible/rest_value/RestValueTest.js:45]: Can you explain why you changed here and the github actions issue

**Review by srknzl — COMMENTED** [review]
Good work in general, I have some questions and feedbacks

File: test/TestUtil.js [hunk 11]
```diff
@@ -125,17 +125,39 @@ exports.markEnterprise = function (_this) {
     }
 };
 
-exports.markServerVersionAtLeast = function (_this, client, expectedVersion) {
-    let actNumber;
+exports.getRandomConnection = function(client) {
+    if (Object.prototype.hasOwnProperty.call(client, 'connectionRegistry')) {
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/TestUtil.js:129]: Can't do that, Eslint prevents such usage

File: test/integration/backward_compatible/ClientFailoverTest.js [hunk 12]
```diff
@@ -73,7 +70,8 @@ describe('ClientFailoverTest - community', function () {
     let cluster2;
 
     before(function () {
-        markCommunity(this);
+        TestUtil.markCommunity(this);
+        TestUtil.markClientVersionAtLeast(this, '4.1');
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientFailoverTest.js:74]: Yes

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js [hunk 13]
```diff
@@ -32,6 +29,10 @@ describe('ClientReconnectAdvancedTest', function () {
     let cluster;
     let client;
 
+    before(function () {
+        TestUtil.markClientVersionAtLeast(this, '4.1');
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js:33]: The mechanism fixed in 4.1

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js [hunk 14]
```diff
@@ -77,19 +78,22 @@ describe('ClientReconnectAdvancedTest', function () {
             },
             lifecycleListeners: [ lifecycleListener ]
         });
-        const connectionManager = client.getConnectionManager();
 
-        await assertTrueEventually(async () => {
-            expect(connectionManager.connectionRegistry.activeConnections).to.have.lengthOf(1);
+        await TestUtil.assertTrueEventually(async () => {
+            expect(client.getCluster().getMembers()).to.have.lengthOf(1);
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js:83]: For this test, it is more than necessary :) We will receive memberlist after connection creation. BTW, this change is not important. We are not testing the connection/member count here. We just try to verify that we reconnect.

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js [hunk 15]
```diff
@@ -77,19 +78,22 @@ describe('ClientReconnectAdvancedTest', function () {
...
         });
 
         await RC.shutdownMember(cluster.id, member.uuid);
         await disconnectedDeferred.promise;
 
+        await TestUtil.assertTrueEventually(async () => {
+            expect(client.getCluster().getMembers()).to.have.lengthOf(1);
+        });
+
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js:92]: Just to make test a bit better

File: test/integration/backward_compatible/ClusterServiceTest.js [hunk 16]
```diff
@@ -50,12 +50,12 @@ describe('ClusterServiceTest', function () {
         const membershipListener = {
             memberAdded: (event) => {
                 expect(event.members.length).to.be.eq(2);
-                expect(client.clusterService.getSize()).to.be.eq(2);
+                expect(client.getCluster().getMembers().length).to.be.eq(2);
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClusterServiceTest.js:53]: No it is not. Check the Cluster interface. ClusterService is internal

File: test/integration/backward_compatible/ClusterServiceTest.js [hunk 17]
```diff
@@ -50,12 +50,12 @@ describe('ClusterServiceTest', function () {
...
                 done();
             }
         };
 
-        client.clusterService.addMembershipListener(membershipListener);
+        client.getCluster().addMembershipListener(membershipListener);
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClusterServiceTest.js:58]: Yes, getCluster is public API, getClusterService is internal

File: test/integration/backward_compatible/MembershipListenerTest.js [hunk 18]
```diff
@@ -46,15 +46,15 @@ describe('MembershipListenerTest', function () {
...
+        client.getCluster().addMembershipListener(membershipListener);
 
         const newMember = await RC.startMember(cluster.id);
         const membershipEvent = await listenerCalledResolver.promise;
 
         expect(membershipEvent.member.address.host).to.equal(newMember.host);
         expect(membershipEvent.member.address.port).to.equal(newMember.port);
         expect(membershipEvent.eventType).to.equal(MemberEvent.ADDED);
-        expect(membershipEvent.members).to.deep.equal(client.clusterService.getMemberList());
+        expect(membershipEvent.members).to.deep.equal(client.getCluster().getMembers());
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/MembershipListenerTest.js:57]: It is private API

File: test/TestUtil.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/TestUtil.js:129]: Yeah you are right

File: test/integration/backward_compatible/README.md [hunk 19]
```diff
@@ -0,0 +1,10 @@
+This folder contains integration tests that should be backward compatible with
+the released 4.x Node.js clients.
+
+One must use mostly public API in the tests located in this folder. If the test
+requires usage of the private API or usage of newly added public API, one must make
+sure it either:
+
+- the private API exists in the all released versions of the client
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/README.md:8]: I am saying the same thing in the entrance paragraph

File: test/integration/backward_compatible/cpsubsystem/AtomicLongTest.js [hunk 20]
```diff
@@ -211,11 +212,13 @@ describe('AtomicLongTest', function () {
     });
 
     it('getAndDecrement: should return old value', async function () {
+        TestUtil.markClientVersionAtLeast(this, '4.1');
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/cpsubsystem/AtomicLongTest.js:215]: No, these methods added later

File: test/integration/backward_compatible/map/MapProxyTest.js (hunk 1 above)
Comment (mdumandag) [reply test/integration/backward_compatible/map/MapProxyTest.js:439]: Member side breaking change. They don't collect per-entry-stats by default anymore

File: test/integration/backward_compatible/map/MapProxyTest.js (hunk 2 above)
Comment (mdumandag) [reply test/integration/backward_compatible/map/MapProxyTest.js:361]: maxIdle parameter support added in 4.1

File: test/integration/backward_compatible/map/NearCachedMapTest.js (hunk 3 above)
Comment (mdumandag) [reply test/integration/backward_compatible/map/NearCachedMapTest.js:249]: same  with maxIdle, setTTL support is added in 4.1

File: test/integration/backward_compatible/nearcache/MigratedDataTest.js (hunk 4 above)
Comment (mdumandag) [reply test/integration/backward_compatible/nearcache/MigratedDataTest.js:38]: If you check the mechanism, it is. We set the whole partition table at once which contains the info related to all partitions. I did this change to make use of the public API

File: test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml (hunk 5 above)
Comment (mdumandag) [reply test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml:1]: Yes, we use this while initializing the cluster in a test

File: test/integration/backward_compatible/serialization/DefaultSerializersLiveTest.js (hunk 6 above)
Comment (mdumandag) [reply test/integration/backward_compatible/serialization/DefaultSerializersLiveTest.js:158]: Yes, they only work in master

File: test/integration/backward_compatible/statistics/StatisticsTest.js (hunk 7 above)
Comment (mdumandag) [reply test/integration/backward_compatible/statistics/StatisticsTest.js:100]: Yes, let's do it later

File: test/integration/javaclasses/PrefixFilter.js (hunk 8 above)
Comment (mdumandag) [reply test/integration/javaclasses/PrefixFilter.js:27]: Well, these files are used in the backward compatible test suite.

File: test/unit/serialization/PortableObjects.js (hunk 9 above)
Comment (mdumandag) [reply test/unit/serialization/PortableObjects.js:30]: These are used in multiple places including backward compatible test suite

File: test/integration/backward_compatible/rest_value/RestValueTest.js (hunk 10 above)
Comment (mdumandag) [reply test/integration/backward_compatible/rest_value/RestValueTest.js:45]: done was getting called more than once. Once for the callback of the itemListener and once for the socket error. If you check the code, what we are checking is unchanged

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js:92]: I guess I need to understand whether member count decreases in case of disconnection. I think member count does not decrease on disconnection. I wonder why we assert member count is 1 three times in this test.

File: test/integration/backward_compatible/ClusterServiceTest.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/integration/backward_compatible/ClusterServiceTest.js:53]: Yeap you right

File: test/TestUtil.js [hunk 21]
```diff
@@ -125,17 +125,39 @@ exports.markEnterprise = function (_this) {
...
+exports.isServerVersionAtLeast = function(client, version) {
+    let actual = BuildInfo.UNKNOWN_VERSION_ID;
     if (process.env['SERVER_VERSION']) {
-        actNumber = BuildInfo.calculateServerVersionFromString(process.env['SERVER_VERSION']);
+        actual = BuildInfo.calculateServerVersionFromString(process.env['SERVER_VERSION']);
     } else if (client != null) {
-        actNumber = client.connectionRegistry.getRandomConnection().getConnectedServerVersion();
-    } else {
-        return;
+        actual = exports.getRandomConnection(client).getConnectedServerVersion();
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/TestUtil.js:141]: If you are OK with the current style, I would like to keep it as it is. This has the same functionality and its aesthetics does not bother me much :)

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js:83]: > BTW, this change is not important. We are not testing the connection/member count here. We just try to verify that we reconnect.

However, the function name is `testConnectionCountAfterClientReconnect` 😄

File: test/integration/backward_compatible/MembershipListenerTest.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/integration/backward_compatible/MembershipListenerTest.js:57]: Right. I did not get clusterService is internal

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js]: Improved the test a little bit. The first member list size check is done to make sure that we are connected. Removed the second check. The third check is replaced by something new that verifies the member list is updated with the new member. Can you check again?

File: test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml (hunk 5 above)
Comment (srknzl) [reply test/integration/backward_compatible/nearcache/hazelcast_eventual_nearcache.xml:1]: Yeah I saw it now. I used find usages feature of jetbrains and it said no usages.

File: test/integration/javaclasses/PrefixFilter.js (hunk 8 above)
Comment (srknzl) [reply test/integration/javaclasses/PrefixFilter.js:27]: Thanks for explaining

File: test/integration/backward_compatible/rest_value/RestValueTest.js (hunk 10 above)
Comment (srknzl) [reply test/integration/backward_compatible/rest_value/RestValueTest.js:45]: What happens when it is called multiple times and why it was "ok" before github actions?

Comment (srknzl) [reply test/integration/backward_compatible/rest_value/RestValueTest.js:45]: Calling done() multiple times throw an error according to mocha docs. But still, why this is related to gh actions?

Comment (mdumandag) [reply test/integration/backward_compatible/rest_value/RestValueTest.js:45]: To be honest, I have no idea why it fails randomly on Github actions :) I tried to reproduce it locally many times with the exact  Node.js, client, and server versions but couldn't succeed. So, I changed the test a little bit. But if you read it, we are testing the same thing

File: test/integration/backward_compatible/ClientReconnectAdvancedTest.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js]: We can compare the uuid's as well. Also, testConnectionCountAfterClientReconnect name does not hold anymore, we can just say `testConnectionAfterClientReconnect` or `testConnectionRestoredAfterClientReconnect`

[the remark this answers is not in the document]
Comment (mdumandag) [reply test/integration/backward_compatible/ClientReconnectAdvancedTest.js]: done. removed the address and port checks since uuid check should be enough

**Comment by mdumandag** [comment]
Thanks a lot for the detailed review
